### DISCUSSION TITLE
Frustrated of FastAPI slow development

### DISCUSSION THREAD

**Discussion by elrik75** [fastapi/fastapi#3970] [Questions] [answered]
There are plenty of good PR asked by a lot of people. No response from @tiangolo 
There are painful bugs like #3665 that we don't know if it will be fixed or not.
Very few commits merged except documentation
Thanks a lot for what you did @tiangolo but you should find additional mainteners to help. Please!

edit: maybe you already search for people, then forget this message :)

**Comment by ghandic** [comment]
@elrik75 I don't think there's much that can happen in this forum (discussions). If you would like to suggest something maybe you could create an issue and add some helpful suggestions to other projects' governance. Possibly following a similar (but less strict) path to the [psf](https://devguide.python.org/coredev/) as that has obviously turned out well with the BFD (Guido) and core contributors. 

I imagine @tiangolo has already put a lot of thought into this already as running these big projects takes a lot of time and investment. Something that often goes unrewarded and prevents you from doing more cool things because you need to worry about maintenance.

**Reply by stephane** [reply]
Example of successful foundation:
- Django, https://www.djangoproject.com/foundation/

Another model is the community-driven organization such as https://www.palletsprojects.com/

**Comment by stephane** [comment]
I was already worried about the situation in June and tiangolo's response increased my concern about the sustainability of the project.

https://twitter.com/sraimbault/status/1403327590039969796

In the meantime (around 5 months), the number of PR and issues has doubled.

BTW: I didn't use SQLModel in my projects because I'm concerned about the maintenance state.

**Reply by ghandic** [reply]
Thanks for adding the link for reference, I think it would be good if there were a roadmap of sorts that detailed the plan for the next X months in terms of governance (eg current plan is everyone can review and @tiangolo can merge) that way it is at least clear as new people submitting PR's keep tagging contributors to merge.

One positive people don't touch on with having sole merger permission is it should make releases more stable, thought out and aligned with the direction of the project.

**Reply by Bluenix2** [reply]
I get where you're coming from, but I don't think this can be considered entirely true.

With only one maintainer you have these issues of slow progress, bugs and issues building up over time etc.

You also talk about an aligned project vision, this is of course true when you only have one maintainer, but is that the best for the project? It can be hard to consider *everything* when making a decision yourself.

**Reply by ccrvlh** [reply]
Same here. Seeing SQLModel issues/PR is a bit scary. Having a one man responsible for everything is usually a high risk scenario, @stephane brought great examples, I wonder what are the plans for the future.

**Reply by hmoravec** [reply]
I'm also starting to be a bit disillusioned. After nice experience in the development phase of a new project with FastAPI, we are facing various problems when moving to production. Starting from the official documentation, where the described DB integration using dependencies leads to deadlocks with more concurrent users (https://github.com/tiangolo/fastapi/issues/3205), reported in May, and ending with various features sort of necessary for production usage where PR has been created a long time ago, but without any reaction, e.g. hiding selected attributes from Swagger, hiding internal documentation from Pydantic models docstrings etc.

**Comment by Bluenix2** [comment]
GitHub recently announced [Merge Queues](https://github.blog/changelog/2021-10-27-pull-request-merge-queue-limited-beta/). This seems perfect for FastAPI where traditionally @tiangolo has been the only maintainer.

From my understanding you "merge" pull requests into a queue, which you can then merge all together. Indicated by the following screenshot it seems like you can allow someone access to add a pull request into a queu *without* allowing them access to merge directly into the main branch..
![Merge Queue options to pick](https://docs.github.com/assets/images/help/pull_requests/merge-queue-options.png)

This means that (potentially) Tiangolo could grant the most active members that he trusts permission to use this feature, and then coordinate these merges into the main branch every now and then.

**Reply by Bluenix2** [reply]
I see a downvote on my comment here, is there any reasoning behind that?

I would like to know what could possibly be the downsides to such a system? Of course it is not a full organization, but this would be a good compromise: more people help keep the project moving while Tiangolo is not forced to give up all control (which I assume he does not want to do).

Also, for those who reacted with a thumbs up.. there's an arrow up to the left you use to upvote!

**Comment by benjamin-kirkbride** [comment]
Some discussion of this here: https://news.ycombinator.com/item?id=29471609

**Comment by benjamin-kirkbride** [comment]
I agree completely with the premise. @tiangolo clearly is very capable and competent at creating tools that are useful and well built, but that is not enough.
 
FastAPI and friends needs an (ideally democratically governed) organization around it to move it forward. Something like Pallets for Flask.

**Comment by acidjunk** [comment]
Is there a vision somewhere for this project? I did find some info in the docs that points to a sponsorship: https://github.com/sponsors/tiangolo

It has a couple of different prices/tiers: but seems to focus on helping customers (which is fine I guess, but this doesn't help in maintaining the issues and open PR's.) For me, most of the sponsorship info, was new and I will discuss it with some of my customers.

I think that a lot of Flask people switched to FastAPI because it's excellent for building Rest API's. Maintaining the code will be more and more important in this stage of the project. So some info about the roadmap and organisation behind FastAPI would be handy.

Probably good to start an organisation + open discussion for this.

**Comment by acidjunk** [comment]
There is a new issue that discusses the future of FastAPI: https://github.com/tiangolo/fastapi/issues/4263

This ticket can be closed?

**Reply by Bluenix2** [reply]
Other way around, that issue can be closed.

Usually with duplicates you keep the oldest one. Besides, GitHub Discussions are a much more appropriate place to hold discussions like this.

**Reply by k4black** [reply]
Sorry, I didn't find this discussion when creating the issue

Let the issue bothers eyes for a while  
I think a yet another link to this discussion is helpful so more people can join

**Reply by benjamin-kirkbride** [reply]
completely agree @k4black, it's been brought up several times before.

**Reply by acidjunk** [reply]
hmm, quite messy forum structure. There is also a comment in this thread that says: please create a ticket.

**Comment by tiangolo** [comment, marked as answer]
Hey! Thanks a lot for your interest.

## Numbers

Let's start with the numbers. I think you might be evaluating the number of issues and PRs with a different/biased point of view, here's how.

Most of the issues are actually **questions**, and **not bugs**. And I use issues for questions because those allow imposing a bit of structure that can help users, me, and any of you here that also help. Additional to that, I don't close issues once I or someone answers them, I feel that would be a bit rude (although maybe practical). I wait for the original poster to close them or I mark them to be closed later by a bot. Many of those questions are already answered and I haven't had the time to mark/close them. But that's what counts for the biggest chunk of the "issues open". Also, some of the issues are just a place to handle some admin, for example, there's an issue for each one of the languages being translated, and I wrote a bot that comments there to let participants to know there's a new translation to review. Those are not many, but still, another chunk of "issues" that are not really issues and are not expected to be "closed".

Now, about PRs, a big chunk of the PRs open are just translations. I can't merge them until I get 2 approving reviews from native speakers. So they just stay there for a while. And I can't force people to come and help reviewing translations for their language.

Then, there are many PRs that I would have an easier time re-implementing them myself than updating the code from the PR, maybe they add extra things not relevant to the current PR, or maybe they implement it in a way inconsistent with the rest of the code, etc. But I try to put an effort into updating the original PR instead of just discarding and re-writing. That also makes the process a bit slower.

## Issues and PRs to handle

It's true I do have a bunch of issues and PRs to review across the projects (not just in FastAPI).

But as I personally review and in most cases fine-tune and update each one of the PRs (if you check the history, almost no PR is directly merged, most of them require updates) it's taking me a bit to handle them all, but I'm on it.

For the same reason, I can't just give permissions to others to simply merge PRs. An important part of what has worked with FastAPI is that I have personally and very carefully handled it all, the best way I can. I can't renounce to that now, at least not yet, and risk the codebase quality. I would probably give permissions to others (and I have done it in the past), after several non-trivial PRs that don't need any change from me. But that doesn't happen frequently. And still, even after some trusted approvals, I would try to review each PR myself before merging.

## Short Term Future

I even changed my working structure to optimize for more open source. I didn't get the chance to work as much as I hoped for last year. I'm finishing a very large migration to FastAPI. This will also result in documentation around that process and how to handle compatibility with other frameworks and async/sync code. But I'm already arranging to compensate it this year soon that I finish that migration, and I'll spend a lot of my working time this next quarter or two on FastAPI and friends.

During this time, I'll also get the chance to see many contributions from others, and who knows, maybe I'll feel more confident about giving extra permissions to some others.

## Conversations like this one

Sadly, new issues and new discussions like this one asking why the other issues are not solved, don't really help, as they just add another issue for me to read and take care of.

It also doesn't help to comment or send me private messages or emails telling me to handle one specific issue. That becomes *another* message I have to handle on top of the thousands I receive from GitHub notifications for issues and PRs.

In the same way, asking me to find new maintainers doesn't really help me, because it just becomes another task you are asking me to carry out. It's also not something that I can myself go out and find. I wouldn't go and ask people to come and help for free, while at the same time I have never seen their work, in particular here, or I have seen only very little. It's also not a matter of people simply volunteering to be maintainers if they have never contributed a non-trivial PR (or very few). None of those options are really enough for me to give them permissions.

But again, as I get the chance to review PRs, maybe I'll find some active contributors with impeccable code and reviews that I would feel confident about giving them more access.

## Recent Work

I'm prioritizing the work that can have the most impact.

Recently I was helping a bit with AnyIO and Trio, as AnyIO is used underneath by FastAPI (and it requires double compatibility with asyncio and Trio) and all that could end up used by you and your projects, even if you didn't know about it. That's not even visible in my projects, but you would still benefit from that work. For example, helping fix support for contextvars when executing synchronous tasks in a thread worker from an async context. You might end up needing all that in your projects, even without knowing, because it could be used by third party libraries you use. And these improvements make sure everything will simply work correctly underneath without you having to worry too much about it.

I was also working on a couple of small things for Pydantic, that would be used by SQLModel and FastAPI.

Before that, I released SQLModel, as I had the code, the ideas, and everything mostly ready, and people seemed to be needing something like it and no one was able to make it, so I spent the time to finish a first version that people could start exploring, before jumping to the next most impactful thing to do.

The same way, I just released Asyncer a couple of days ago, to help the projects that need to work with concurrent code and that need to mix async and sync code, which I would think is a very strong use case for FastAPI.

## Roadmap

I have several things that I want to work on, but my roadmap is flexible, as while working on one thing I could end up finding that there's something with high impact that I could do, and if it's worth it, I might just do that instead of a previously fixed plan that was made based on suppositions.

I currently don't have investors or similar, so I don't have to force myself to meet some product or marketing deadlines, or to figure out how to monetize something, etc. I can focus on whatever seems like what would have the most impact. Even if that means working on a third party open source library that isn't even mine, but still feels like the right thing to do. I can do that now. That's also why the roadmap is not very fixed.

As an example, I want it to be easier for people to work with async code. To run concurrent tasks, to be able to mix sync code with async code but reducing all the confusion around threads, thread workers, event loops, contextvars, thread locals, etc. That includes a lot of documentation I want to write. And it also ended up including building Asyncer, to be able to have a good developer experience with autocompletion, inline type errors, etc.

All that work on Asyncer started as PRs to Trio and AnyIO, and then in Asyncer itself, which is this tiny library to let me (and you if you want) use that API design, while those PRs are evaluated in Trio and AnyIO themselves. But the realization that it would make sense to make it its own tiny library, even if temporary, was last month, and then I finished it (almost all of it) during the break.

All that to say that I feel it's an example of the benefits of my flexible roadmap and focusing on the resulting impact.

## Come and Help me Maintain

The word "maintain" and "maintainer" probably means different things to each one of you, and each one of you probably measures it in different ways.

If you want to see faster progress, there are several ways to help, they are all documented here: https://fastapi.tiangolo.com/help-fastapi/

One of the things that consume time the most for a maintainer is handling issues created by others.

Many issues are incomplete questions. If you can go and check them, and if you can write a simple, minimal, self-contained example that shows the actual issue (or shows how to solve the user's question), that's normally the biggest chunk of time and you would be saving me from doing it myself. Then, **you would be maintaining FastAPI**. If it's actually a bug, I can probably reuse your example as a test case, and that's probably the biggest part of the work.

If you go and help them and they close those issues, that's a lot of minutes (in some cases hours) that you save me, and that I can then dedicate to review the other issues and PRs. That's work that a **maintainer** has to do. Wanna do it? I would be most grateful! Any of you can do it.

You can also review PRs yourself. But have in mind that adding a comment of "I have this problem too, merge this soon" doesn't really help if you are not really reviewing the code, trying it in your project, and confirming that it indeed solves your use case. And if you did all that, please mention it.

In some way, here's me **asking you all to be maintainers for FastAPI**.

## FastAPI People

A lot of this work is already done by a bunch of people. Not many, but a few very dedicated and helpful volunteer people.

For example, @Kludex is one of the most helpful people here, always helping others, and always with a supportive and friendly attitude.

That's why I created FastAPI People, and the FastAPI Experts. Those are the people that have helped maintain FastAPI, helping others with their questions: https://fastapi.tiangolo.com/fastapi-people/#experts

Do you have open issues? Go check it. Are they solved? Then please close them. Help me maintain this.

On the other hand, overly critical, aggressive, and demanding comments like those from @stephane only drain me, make me sad, and take out the energy I should be spending in working on this instead of arguing. At the same time, none of those users show up in FastAPI People, which means that none of them are helping FastAPI, helping maintain it, answering questions in issues, reviewing PRs, etc.

But for the rest of you, thanks a lot for your support and understanding. 🙇

**Reply by Bluenix2** [reply]
>Most of the issues are actually **questions**, and not **bugs**.

For anybody interested - who's maybe not a search-god - this can easily be proven correct by searching for [`is:open is:issue -label:question`](https://github.com/tiangolo/fastapi/issues?q=is%3Aopen+is%3Aissue+-label%3Aquestion+) on the GitHub issues.

As I am writing this message, that filters out roughly 860 issues to **only about 200** (rounded to the nearest "tenth").

> Now, about PRs, a big chunk of the PRs open are just translations. I can't merge them until I get 2 approving reviews from native speakers. So they just stay there for a while. And I can't force people to come and help reviewing translations for their language.

This we can also search our way to, an example would be going to the [5th page right now](https://github.com/tiangolo/fastapi/pulls?page=5&q=is%3Apr+is%3Aopen) which shows a giant wall of Chinese translations by @jaystone776. Again, we can filter out all translations using [`is:open is:pr -label:lang-all`](https://github.com/tiangolo/fastapi/pulls?q=is%3Aopen+is%3Apr+-label%3Alang-all+) and (as of writing this) we go from 440 pull requests to 180 (also rounded). 

That said if we search for non-triaged pull requests with [`is:open is:pr no:label`](https://github.com/tiangolo/fastapi/pulls?q=is%3Aopen+is%3Apr+no%3Alabel), it results in 170 (rounded) PRs (including a few not correctly labeled translation PRs). This does suggest that all PRs that have been given labels have also been merged.

---

I do really want to say that I truly do believe that most people - me included - have been commenting on this discussion with the best intent. There are a lot of people who really want to see this project succeed. Commenting on this discussion is not meant to by any means bash you personally or your work, rather this is more about the fact that there's more issues and PRs than you have time for - and that's just how it is sometimes.

Now it is true that issues and discussions like these do take away from time you would spend on issues and PRs, but the point of them are to have a net-increase in productivity.

---

> Additional to that, I don't close issues once I or someone answers them, I feel that would be a bit rude (although maybe practical). I wait for the original poster to close them or I mark them to be closed later by a bot. Many of those questions are already answered and I haven't had the time to mark/close them

Perhaps this is not the best for the project, Discussions seem like a much better place to have questions go - and exactly how it appears that the feature is meant to be used. I will quicky mention that the fact that this repository is on your personal account puts a lot of restrictions on permissions you can allow and features you can use. [Merge Queues I mentioned above](https://github.com/tiangolo/fastapi/discussions/3970#discussioncomment-1569451) as well as [fine grained repository access](https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-roles-for-an-organization) both seem to be limited to organizations sadly. If that wasn't the case, I would genuinely suggest adding @Kludex and other people as triagers having access to add/remove labels - that can do a lot for this project.

> I'm prioritizing the work that can have the most impact.

That is awesome and I have seen some of your PRs when browsing AnyIO as well as Trio - after all, it is your time. You spend it however you like (although I get that there surely are people out there on the internet criticizing you for that), you could spend none on FastAPI - but then, for the project's sake, let someone spend time on it.

Thank you for taking the time to type out such a lengthy reply.

**Reply by tiangolo** [reply]
Thanks for your comments, help, and support @Bluenix2! :cake:

**Reply by stephane** [reply]
Thank you @tiangolo for this very detailed answer on the governance you want for FastAPI.

Choosing a framework can be a big risk for a company. Over the months, I have been reading worried comments hoping for a reaction from you and you finally provide an answer.

I'm disappointed that you choose to **name and shame**.

In this thread, I expressed my concerns about the [bus factor](https://en.wikipedia.org/wiki/Bus_factor), I don't think it requires this attitude.
However, I wrote a comment in another thread: https://github.com/tiangolo/fastapi/issues/4263#issuecomment-1007273915

Please, accept my apologies for the last line of that comment.

You can find my contributions to FastAPI ecosystem here:
- https://github.com/fastapi-users/fastapi-users/
- https://github.com/yezz123/authx#readme

I wish the FastAPI project the best 🚀

**Reply by benjamin-kirkbride** [reply]
@tiangolo thank you for your response.

To me, the biggest takeaway here that the bus factor of one is going remain indefinitely. Thank you for the clarification on that.

I would like to see that information added to all the places you state that FastAPI is "production ready" as that is info that I would have loved to have when I decided to use it in production, (though it definitely could be argued that I should have not have used a [ZeroVer](https://0ver.org/) project regardless, live and learn).

**Reply by four43** [reply]
So we need to hope that God's gift to Python doesn't get smoked by a bus or loose interest? That seems like a big wager for an awesomely popular library like this. I get this is your pride and joy, and you don't owe us anything. But it's crazy to expect one dude to keep up with all of this. And it's obviously not working. Not just for fastapi but starlette too. It's a lot, they've grown a lot. Let us help. 

This seems like a hard fork and restructure sort of situation in order to save any sort of confidence in this little ecosystem.

**Reply by adriangb** [reply]
Hey @four43 I am not the person being referenced, but if I was, I would find your comment unnecessarily aggressive and personal for what should be a professional discussion. The point can be (and has already in this thread) made with more compassion for the human on the other side. Just letting you know that it may not read as you intended.

> it's crazy to expect one dude to keep up with all of this. And it's obviously not working. Not just for fastapi but starlette too

And just FYI Starlette is not "one dude": there are currently 35 maintainers and Tom Christie, the creator of the project, has done a lot of work recently to derisk from the bus factor of 1. I'm not saying that's the better path forward, just that your statement is not accurate.

**Reply by four43** [reply]
This is quickly branching into commentary on the industry as a whole - but there are many large houses of cards relying on projects like this. Which is great as a project owner - that's a successful project. But there are big backlogs of things that need to get done and pretty substantial bugs that hold it back. Maybe that's good - it keeps mess out of the code base. A swift knife to cut through bullshit is good (look at the Linux Kernel) but it still warrants concern that there is one person who's quite hesitant to hand over any sort of power for folks eager to help. The biggest stink of a overwhelmed code base are old unreviewed PRs. Issues are whatever - lots of newer folks posting StackOverflow like questions there, but PRs are people who have taken the time to help, want to help, and are still stuck. There's other smart devs out there, thinking you're the only one just a jr. mistake.

Good to hear starlette is starting to make some more progress. That seemed to be where the core of the good ideas lie. They haven't closed out some of the tickets I was watching last I checked, but maybe they're getting there.

**Reply by lovelydinosaur** [reply]
Heya folks. 👋🏼

Some possible actions which may or may not be helpful?...

> Most of the issues are actually questions, and not bugs. And I use issues for questions because those allow imposing a bit of structure

Would you consider switching on a Stale bot? Exempt anything that's properly labelled as a non-question, and allow the bot to close questions off for you. [Example](https://github.com/encode/httpcore/blob/master/.github/stale.yml), [Docs](https://github.com/apps/stale).

Alternately you could migrate them to GitHub "discussions"? I realise this might not be an option for you, as I think you've got some workflow/tooling around issues and the rather neat [FastAPI people](https://fastapi.tiangolo.com/fastapi-people/) docs page.

>  I can't just give permissions to others to simply merge PRs.

Blindly adding merge permissions and on-boarding extra folks isn't in itself a solution to anything much at all. It's very easy to end up *creating* extra work, code churn and project risk, rather than *reducing* it.

From my perspective something that might benefit FastAPI at the moment could be a very clear feature-pause policy.

Have a blanket policy of "no" to any code refactorings, tooling refactorings, and new features, at least until the issue and PR backlog is in better shape.

You could then start by granting a few write-level permissions collaborators, but under the guidance that *for now* those permissions are *only* used to close off issues, or to merge documentation & translation PRs.

That'd give you a comfortable route to *starting* to open up to a wider maintenance team, perhaps?

> A big chunk of the PRs open are just translations. I can't merge them until I get 2 approving reviews from native speakers.

Perhaps that requirement is too much? I think even pretty relaxed policies like "no native language reviewer required" could probably be mostly okay. May well be a better trade-off between managing backlog vs. ensuring translation quality, until/unless you see evidence to the contrary.

**Reply by rogerbrinkmann** [reply]
> > A big chunk of the PRs open are just translations. I can't merge them until I get 2 approving reviews from native speakers.
> 
> Perhaps that requirement is too much? I think even pretty relaxed policies like "no native language reviewer required" could probably be mostly okay. May well be a better trade-off between managing backlog vs. ensuring translation quality, until/unless you see evidence to the contrary.

I think, that's a really good proposal. Some translations seem to be more than 6 month old and the original authors might have forgotten about them or lost interest because there is just no one to review their work. I for my part would love to contribute more German translations, but seeing old forgotten pull requests, where the englisch version is already far ahead is quite discouraging.
I don't think the quality would suffer much by having only one reviewer. And there is always the option to fix mistakes afterwards. As long as people see things moving head.

**Reply by nuao** [reply]
@tiangolo 

You’re talking a lot about what *could be* and how you imagine things *should be*. 
But I want to take a moment and look at how things are right now.

You are asking people who want to contribute to help out with the open issues.
That is a great idea, and you are right in that a good chunk of them are simply questions.
The part that doesn’t work out though is the one where you’re assuming that if those questions will be answered, their authors will close them.
It may happen in some cases, but the reality is, most of these issues are stale. 
You won’t get any feedback from the original author, the question is simply not relevant anymore or they won’t come back and close it themselves. I know you want them to, and I agree that it would be great if they did, but it just won't happen. It’s like this in most major open source project, and FastAPI is no exception here. 
So most of these issues will stay open, no matter how much effort the community puts into them.

And seeing this state just won’t attract people to contribute. You’re saying that you will close the issue when they’re resolved, but actually, this doesn’t happen. I’m sure there are multiple people here that would gladly take it upon themselves to work through all the issues if they knew that they’d be closed / marked as resolved in a timely manner, or if they could do so themselves, but this is simply not the case. 

Your focus is on churning out cool and interesting new libraries. That’s totally fine, and very much appreciated. 
You want to keep doing this stuff.
But I think you’ll have to acknowledge that it takes more time and effort than that of a single developer for whom it isn’t even a top priority to maintain a project like FastAPI. 

You’re going to have to delegate some of the things. And you’ll have to trust that whoever you’re giving this power to is capable of doing a job as good as you and that’s in the best interest for the project. Because if you don’t, there’s very little point in delegating the work at all, since you’ll have to review and look at in in detail yourself anyway.

If you want good, high quality PRs, people have to see that those PRs actually stand the chance of getting merged. 
This also isn’t happening. There are multiple open PRs that, without input from you detailing why they’re not ready to be merged, can only be considered high quality and merge-able, but have gone stale now because of the lack of aforementioned input. 
I get that you only want the best for your project, but with workflows like this, those PRs will never come. 

And it’s only understandable. 
Would you sink hours or days into fixing issues and creating PRs on a project where getting feedback, let alone approval could take months, if it happens at all? Most people won’t. 
And so a feedback loop is created where you are looking for those good PRs and unicorn candidates for possible maintainers with the actual power to do something, but without them being able to actually contribute in a meaningful way to this project, how could you possibly find them?

You are asking the community to step up and do stuff, but at the same time you’re not really giving them the opportunity to do so.
If it’s your wish to remain the only person with any form of write access to this repo, you’ll have to be responsive. 
If you want the community to be active and do stuff, you have to be reachable. Otherwise it can’t work. 
**If your feedback and reviews are needed to get something merged and if you want the community to fix issues and come up with PRs, you’ll have to provide those in a timely manner.**

Don’t get me wrong, I’m not saying any of this is your obligation. It isn’t. I’m just trying to point out that the things you want / expectations you have are slightly contradictory. 

Basically there are three scenarios:

- The community cannot work independently from you: You need to be involved regularly
- You are not involved regularly: The community will need more freedom to move things forward
- The community cannot work independently from you and you’re not regularly involved

The first two are viable. The last one isn’t. But that’s the one currently in place.

**Reply by acidjunk** [reply]
Maybe a community driven online place for questions about fastapi would be helping in closing a lot of the issues.

**Reply by havardthom** [reply]
> Maybe a community driven online place for questions about fastapi would be helping in closing a lot of the issues.

FastAPI has an active discord server and gitter chat: https://fastapi.tiangolo.com/help-fastapi/#join-the-chat

**Reply by tiangolo** [reply]
Thanks for the input everyone!

I just updated the docs about helping FastAPI, there's a new section [**Help Maintain FastAPI**](https://fastapi.tiangolo.com/help-fastapi/#help-maintain-fastapi).

The biggest work maintaining FastAPI can be done by any and all of you here. And that work is what demands the most time.

There are many more details about how to help maintain FastAPI, the specific tasks, and what **YOU** can do, right now.

So, please, come and help me maintain FastAPI! :muscle: :nerd_face:

**Comment by matthewmercuri** [comment]
I hope everyone is doing well. I just want to share some of my own thoughts after reading through this thread. I have used FastAPI for a couple of projects and definitely hope to use it more. I really do like this project and want it to succeed. I think it goes without saying that everyone in this thread feels the same way.

I specifically want to focus on one pain point I currently have with FastAPI and its development: **optics**. I am not the most experienced developer, but I think a key component for any new project to take off is community sentiment. More specifically, it needs to viewed favourably by the community of developers that may eventually use the project. Recently there was a thread on r/python titled _"Django or Flask and why?"_ After scrolling though the comments, I was excited to see someone say FastAPI! It really does seem that those who use the project instantly become advocates for it. Now the unfortunate part, a reply to that polarizing suggestion to use FastAPI:

> I wish people would stop recommending this. Or that people knew how to vet an open source project. Because this project isn’t actively maintained, and its developer has shown an.. interesting approach to dealing with the open source community.
> 
> His last statement, after months of inactivity and being unreachable (while being the sole maintainer of FastAPI), was that he does not plan or wish to enable people other than him to actively maintain it. People tried to reach out to him about the maintenance issue for a while, but this was all they got, and after that, he disappeared again.
> 
> (Same is true for all his other projects. He creates them, maintains them for a couple of weeks / months, heavily adversities them and gets them sponsors by using his platform and reach, and then pretty much abandons them)
> 
> So while FastAPI may have had some good ideas, it doesn’t seem like the project has a future.

Obviously, this commenter was ill-informed. That should definitely be obvious at this point. In fact, a reply to the comment pointed this out and even mentioned this exact GH discussion. To me, this comment is strictly a symptom of the aforementioned **optics** pain point. In this case, a user that would have otherwise liked to try out FastAPI after reading the Reddit thread may have been instantly put off because others have replied declaring the project dead on arrival. Presumably these responders have viewed the GH page some time before, saw the overwhelming amount of issues and open PRs, and thought "oh well." Even if they were more critical at assessing the project they would notice things like:
- infrequent merges
- non-shrinking amount of PRs and issues
- a barrage of new issues that are not even tagged

I do not wish to come across as potentially having all the answers, nor someone beating the same drum as others above. I merely would like to present that the Python community does not see FastAPI in the light I believe it should. We have to be honest and realize this likely comes with a great cost. There are a myriad of potential users and contributors that are not gravitating towards Fast API. Whether that is a function of the issue count, PR count, or community influence is irrelevant. There needs to be something put in place to ensure the community knows this project is not only active, but also fantastic! I have no doubt that with some of the smartest developers around lurking here and a very intelligent, skillful creator, this problem can be remedied.

**N.B.** Please do not try to find the commenter on Reddit and harass them. It is not a bad take and solely indicative of a problem we can solve.

**Reply by four43** [reply]
Well said.

**Reply by benjamin-kirkbride** [reply]
Yeah, this is a great point. Every Hacker News and Reddit thread I have seen that mentions FastAPI for the last year or so has multiple people pointing out that the projects are unmaintained, and since Tiangolo responded to that criticism here and in a few other places, those criticisms now typically include that and that he has expressed negative interest in open governance of the ecosystem's projects and that he feels that the bigger problem is the criticism itself.

Fair or not, this is the perception that is rapidly spreading.

**Comment by ddahan** [comment]
Hi everyone, new fastAPI developer here, finding this thread kind of randomly.
As the messages are 2 years old now, how would you describe now in september 2024, the evolution of all of this?
The current ratio of open/closed PR/issues seam pretty decent right now.

**Reply by elrik75** [reply]
Hi @ddahan. 
2 years later, I can say things are going better. For my usage, I don't have huge issues with FastAPI.
I created this topic because the fastAPI author @tiangolo was very alone on the project and I was afraid about long term support but now we can see more contributors.
Anyway, I think FastAPI is now the best python framework for a ReST API. It's fast, well documented and well supported. Enjoy!
